### PR TITLE
Seq

### PR DESCRIPTION
**Pull request by WangBoxue**
Add functions below:
newSeq, removeIf, getIf, get, repeat, at, compact, count

### CODE REVIEW

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 1]
```diff
@@ -389,6 +390,32 @@ default int findLastIndex(Predicate<T> condition) {
     }
 
     /**
+     * Create a new seq with size null elements
+     */
+    static <T> Seq<T> newSeq(int size) {
+        return new SeqImpl<>(size);
+    }
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: Please follow the original documentation if you can:

```
    /**
     * Constructs an empty list with the specified initial capacity.
     *
     * @param  initialCapacity  the initial capacity of the list
     * @throws IllegalArgumentException if the specified initial capacity
     *         is negative
     */
    public ArrayList(int initialCapacity) {
```

You don't need to spend time to write your own, which is not so better than the original.

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 2]
```diff
@@ -389,6 +390,32 @@ default int findLastIndex(Predicate<T> condition) {
...
+    /**
+     * Create a new seq with size defaultValue elements
+     */
+    static <T> Seq<T> newSeq(int size, T defaultValue) {
+        return new SeqImpl<>(size, defaultValue);
+    }
+
+    /**
+     * Create a new seq with size elements, each of elements is generated by a {@link Function}.
+     */
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: Take care of English, write good documentation. For example:

What do you mean by "size elements"?

"each of elements is generated by a {@link Function}". Users will ask "what is the input of this function?", index or anything else?

The original Ruby documentation:

> In the last form, an array of the given size is created. Each element in this array is created by passing the element’s index to the given block and storing the return value.

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 3]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
...
+        }
+        return this;
+    }
+
+    /**
+     * Get elements which satisfy the condition, resulting a new seq without changing the original one.
+     *
+     * @return The new seq
+     */
+    default Seq<T> getIf(Predicate<T> condition) {
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: We'd better unify the name, how about `filter`? I think Java 8 already use this word.

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 4]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
...
+        return this.get(i);
+    }
+
+    /**
+     * Returns a new {@link Seq} built by concatenating the times copies of self.
+     * @return the new seq
+     */
+    default Seq<T> repeat(int times) {
+        if (times < 0)
+            throw new IllegalArgumentException("times must be a positive number.");
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: Add exception to doc

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 5]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
...
+                times--;
+            }
+        }
+        return this;
+    }
+
+    /**
+     * Returns the element at index. A negative index counts from the end of self.
+     */
+    default T at(int index) {
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: How about overriding `get` method to achieve the same function?

Which do you think better?

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 6]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
...
+    /**
+     * Returns a copy of self with all null elements removed.
+     * @return the new seq
+     */
+    default Seq<T> compact(){
+        return this.getIf(e -> e != null);
+    }
+
+    /**
+     * Returns self with all nil elements removed.
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: `nil` -> `null`

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 7]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
...
+     */
+    default Seq<T> compact$(){
+        removeIf(e -> e == null);
+        return this;
+    }
+
+    /**
+     * Returns the number of elements.
+     */
+    default int count(T element) {
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: What is the difference with `size`?

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 8]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
...
+        });
+        return copy;
+    }
+
+    /**
+     * Remove elements which satisfy the condition at the original seq.
+     *
+     * @return The seq itself after the change
+     */
+    default Seq<T> removeIf$(BiPredicate<T, Integer> condition) {
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: Consider to use `reject`. Ruby has the style to make many names on the same function. But I think we don't need to follow that in Java. So I think to make the best name is very important

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 9]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
     default Seq<T> subSeq(int fromIndex, int toIndex) {
         return new SeqImpl<>(this.subList(fromIndex, toIndex));
     }
+
+    /**
+     * Remove elements which satisfy the condition, resulting a new seq without changing the original one.
+     *
+     * @return The new seq after the change
+     */
+    default Seq<T> removeIf(BiPredicate<T, Integer> condition) {
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: To keep consistency with other method styles, I think we need a lambda for only the element without index?

File: src/test/java/com/worksap/fig/SeqTest.java [hunk 10]
```diff
@@ -212,4 +213,106 @@ public void testConcatPreConcat() {
...
+
+    }
+
+    @Test
+    public void testGet() {
+        Seq<Integer> seq = Seq.of(1, 2, 3);
+        assertEquals(new Integer(1), seq.get(0));
+        assertEquals(new Integer(100), seq.get(-1, 100));
+        assertEquals(new Integer(100), seq.get(4, 100));
+    }
```
Comment (LeonLiuY) [inline src/test/java/com/worksap/fig/SeqTest.java]: You're missing the test for "legal index and different with default value"

File: src/test/java/com/worksap/fig/SeqTest.java [hunk 11]
```diff
@@ -212,4 +213,106 @@ public void testConcatPreConcat() {
...
+        assertEquals(2, seq.count(e -> e == null));
+        assertEquals(3, seq.count(e -> e != null && e > 1));
+        assertEquals(2, seq.count((e, i) -> e != null && e > 1 && i < seq.size() - 1));
+    }
+    @Test
+    public void test() {
+        Seq<Integer> seq = Seq.of(1);
+        seq.clear();
+
+    }
```
Comment (LeonLiuY) [inline src/test/java/com/worksap/fig/SeqTest.java]: I guess this method is for debug? If so we should remove it

File: src/main/java/com/worksap/fig/lang/Seq.java [hunk 12]
```diff
@@ -404,4 +431,189 @@ default int findLastIndex(Predicate<T> condition) {
...
+        return copy;
+    }
+
+    /**
+     * Get the element with index. If the index is illegal, return a default value
+     * @param i index
+     * @param defaultValue default value
+     * @return the element or default value
+     */
+    default T get(int i, T defaultValue) {
```
Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: I guess this method in Ruby is called `fetch`,  do you think `get` is better?

File: src/main/java/com/worksap/fig/lang/Seq.java (hunk 7 above)
Comment (WangBoxue) [inline src/main/java/com/worksap/fig/lang/Seq.java]: count() is used to get the number of one element or elements that satisfy the condition.
size() is used to get the number of all elements.

Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: I understand. Do you think the doc can explain this meaning?

File: src/main/java/com/worksap/fig/lang/Seq.java (hunk 1 above)
Comment (WangBoxue) [inline src/main/java/com/worksap/fig/lang/Seq.java]: What do you mean 'your own'? The doc or the function?

Comment (LeonLiuY) [inline src/main/java/com/worksap/fig/lang/Seq.java]: I mean the doc

File: src/main/java/com/worksap/fig/lang/Seq.java (hunk 12 above)
Comment (WangBoxue) [inline src/main/java/com/worksap/fig/lang/Seq.java]: Yes, it's called fetch in Ruby. Ruby doesn't have a function named get. But Java has one function called get. In order to keep consistency, I use the name of get.

File: src/main/java/com/worksap/fig/lang/Seq.java (hunk 7 above)
Comment (WangBoxue) [inline src/main/java/com/worksap/fig/lang/Seq.java]: No, I have changed it.

File: src/main/java/com/worksap/fig/lang/Seq.java (hunk 9 above)
Comment (WangBoxue) [inline src/main/java/com/worksap/fig/lang/Seq.java]: I think we should provide both. Just like `forEachWithIndex` and `forEach`.

**Comment by WangBoxue** [comment]
@liuyang1204 I have revised the docs, function names and some implementations. Please check.

**Comment by LeonLiuY** [comment]
Great!

**Pull request merged**